### ISSUE TITLE
SOAP plugin issue

### ISSUE THREAD

**Issue by bhavikas-inx** [outofcoffee/imposter#281]
Hello,
Unable to start imposter after installing mock-soap plugin with configurations.
I am getting this kind of errors.

`DEBU[0000] plugin store-redis version 3.5.1 is installed 
DEBU[0000] plugin mock-soap version 3.5.1 is installed  
INFO[0000] starting mock engine on port 9090 - press ctrl+c to stop 
DEBU[0000] engine image '3.5.1' already present         
13:30:36 INFO  i.g.i.Imposter - Starting mock engine 3.5.1
13:30:37 DEBUG i.g.i.c.u.ConfigUtil - Loading configuration file: /opt/imposter/config/mock-config.yaml
13:30:38 ERROR i.v.c.i.l.c.VertxIsolatedDeployer - Failed in deploying verticle
java.lang.RuntimeException: Failed to register plugin: soap
	at io.gatehill.imposter.plugin.DynamicPluginDiscoveryStrategyImpl.registerPluginClass(DynamicPluginDiscoveryStrategyImpl.kt:136) ~[imposter-config-dynamic-3.5.1.jar:?]
	at io.gatehill.imposter.plugin.DynamicPluginDiscoveryStrategyImpl.processPlugin(DynamicPluginDiscoveryStrategyImpl.kt:108) ~[imposter-config-dynamic-3.5.1.jar:?]
	at io.gatehill.imposter.plugin.DynamicPluginDiscoveryStrategyImpl.preparePluginsFromConfig(DynamicPluginDiscoveryStrategyImpl.kt:87) ~[imposter-config-dynamic-3.5.1.jar:?]
	at io.gatehill.imposter.plugin.PluginManagerImpl.preparePluginsFromConfig(PluginManagerImpl.kt:67) ~[imposter-config-3.5.1.jar:?]
	at io.gatehill.imposter.Imposter$start$1.invokeSuspend(Imposter.kt:114) ~[imposter-engine-3.5.1.jar:?]
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33) ~[kotlin-stdlib-1.7.10.jar:1.7.10-release-333(1.7.10)]
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106) ~[kotlinx-coroutines-core-jvm-1.6.4.jar:?]
	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:570) ~[kotlinx-coroutines-core-jvm-1.6.4.jar:?]
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:750) ~[kotlinx-coroutines-core-jvm-1.6.4.jar:?]
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:677) ~[kotlinx-coroutines-core-jvm-1.6.4.jar:?]
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:664) ~[kotlinx-coroutines-core-jvm-1.6.4.jar:?]
Caused by: java.lang.ClassNotFoundException: soap
	at jdk.internal.loader.BuiltinClassLoader.loadClass(Unknown Source) ~[?:?]
	at jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(Unknown Source) ~[?:?]
	at java.lang.ClassLoader.loadClass(Unknown Source) ~[?:?]
	at io.gatehill.imposter.util.ClassLoaderUtil.loadClass(ClassLoaderUtil.kt:98) ~[imposter-config-3.5.1.jar:?]
	at io.gatehill.imposter.plugin.DynamicPluginDiscoveryStrategyImpl.registerPluginClass(DynamicPluginDiscoveryStrategyImpl.kt:126) ~[imposter-config-dynamic-3.5.1.jar:?]
	... 10 more
FATA[0030] timed out waiting for status endpoint to return HTTP 200 at http://localhost:9090/system/status `

**Comment by outofcoffee** [comment]
Hi @bhavikas-inx, thanks for the bug report. May I ask how you are running Imposter? It looks like you're using the CLI. If so, would you mind running:

```
imposter version
```

and, if you're running v0.23.x or later of the CLI:

```
imposter plugin list
```

...otherwise, please could you run:

```
export IMPOSTER_VERSION="3.5.1"
ls -l $HOME/.imposter/plugins/${IMPOSTER_VERSION}
```

**Comment by outofcoffee** [comment]
For info, here is my output:

```
$ imposter plugin list

|   TYPE    | VERSION |
|-----------|---------|
| mock-soap | 3.5.1   |
```

and:
```
$ export IMPOSTER_VERSION="3.5.1"
$ ls -l $HOME/.imposter/plugins/${IMPOSTER_VERSION}

total 6200
-rw-r--r--  1 pete  staff  2270690  2 Jan 14:29 imposter-plugin-mock-soap.jar
```

When I start Imposter with the CLI I get:

```
$  imposter up

INFO[0000] starting mock engine on port 8080 - press ctrl+c to stop
DEBU[0000] engine image '3.5.1' already present
14:36:13 INFO  i.g.i.Imposter - Starting mock engine 3.5.1
14:36:14 DEBUG i.g.i.c.u.ConfigUtil - Loading configuration file: /opt/imposter/config/imposter-config.yaml
14:36:15 DEBUG i.g.i.p.PluginManager - Loaded 7 plugin(s): [js-detector, store-detector, meta-detector, js-nashorn-standalone, store-inmem, config-detector, soap]
14:36:15 DEBUG i.g.i.p.s.p.VersionAwareWsdlParser - Using WSDL parser: V1 for: /opt/imposter/config/service.wsdl
14:36:15 DEBUG i.g.i.p.s.p.Wsdl1Parser - Discovered 1 schema(s) for WSDL: /opt/imposter/config/service.wsdl
14:36:15 DEBUG i.g.i.p.s.SoapPluginImpl - Adding mock endpoint: SoapBinding -> SOAP /pets/
14:36:16 INFO  i.g.i.Imposter - Mock engine up and running on http://localhost:8080
```

**Comment by bhavikas-inx** [comment]
Hello, Yes I have installed imposter CLI using Homebrew. It is still not working. 
Do I miss anything?
I am trying to implement the SOAP simple imposter example. 
[https://github.com/outofcoffee/imposter/tree/main/examples/soap/simple](https://github.com/outofcoffee/imposter/tree/main/examples/soap/simple)

`root087@root087:~/BhavikaSoni/Projects/imposter$ imposter up -p 9090 example/soap/simple
DEBU[0000] plugin store-redis version 3.5.1 is installed 
DEBU[0000] plugin mock-soap version 3.5.1 is installed  
INFO[0000] starting mock engine on port 9090 - press ctrl+c to stop 
INFO[0000] pulling '3.5.1' engine image                 
05:58:13 INFO  i.g.i.Imposter - Starting mock engine 3.5.1
05:58:14 DEBUG i.g.i.c.u.ConfigUtil - Loading configuration file: /opt/imposter/config/imposter-config.yaml
05:58:15 ERROR i.v.c.i.l.c.VertxIsolatedDeployer - Failed in deploying verticle
java.lang.RuntimeException: Failed to register plugin: soap
	at io.gatehill.imposter.plugin.DynamicPluginDiscoveryStrategyImpl.registerPluginClass(DynamicPluginDiscoveryStrategyImpl.kt:136) ~[imposter-config-dynamic-3.5.1.jar:?]
	at io.gatehill.imposter.plugin.DynamicPluginDiscoveryStrategyImpl.processPlugin(DynamicPluginDiscoveryStrategyImpl.kt:108) ~[imposter-config-dynamic-3.5.1.jar:?]
	at io.gatehill.imposter.plugin.DynamicPluginDiscoveryStrategyImpl.preparePluginsFromConfig(DynamicPluginDiscoveryStrategyImpl.kt:87) ~[imposter-config-dynamic-3.5.1.jar:?]
	at io.gatehill.imposter.plugin.PluginManagerImpl.preparePluginsFromConfig(PluginManagerImpl.kt:67) ~[imposter-config-3.5.1.jar:?]
	at io.gatehill.imposter.Imposter$start$1.invokeSuspend(Imposter.kt:114) ~[imposter-engine-3.5.1.jar:?]
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33) ~[kotlin-stdlib-1.7.10.jar:1.7.10-release-333(1.7.10)]
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106) ~[kotlinx-coroutines-core-jvm-1.6.4.jar:?]
	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:570) ~[kotlinx-coroutines-core-jvm-1.6.4.jar:?]
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:750) ~[kotlinx-coroutines-core-jvm-1.6.4.jar:?]
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:677) ~[kotlinx-coroutines-core-jvm-1.6.4.jar:?]
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:664) ~[kotlinx-coroutines-core-jvm-1.6.4.jar:?]
Caused by: java.lang.ClassNotFoundException: soap
	at jdk.internal.loader.BuiltinClassLoader.loadClass(Unknown Source) ~[?:?]
	at jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(Unknown Source) ~[?:?]
	at java.lang.ClassLoader.loadClass(Unknown Source) ~[?:?]
	at io.gatehill.imposter.util.ClassLoaderUtil.loadClass(ClassLoaderUtil.kt:98) ~[imposter-config-3.5.1.jar:?]
	at io.gatehill.imposter.plugin.DynamicPluginDiscoveryStrategyImpl.registerPluginClass(DynamicPluginDiscoveryStrategyImpl.kt:126) ~[imposter-config-dynamic-3.5.1.jar:?]
	... 10 more
FATA[0058] timed out waiting for status endpoint to return HTTP 200 at http://localhost:9090/system/status 
`
`root087@root087:~/BhavikaSoni/Projects/imposter$ imposter version
- imposter-cli 0.24.1
- imposter-engine 3.5.1
- engine-output 3.5.1
`

`root087@root087:~/BhavikaSoni/Projects/imposter$ imposter plugin list
|    TYPE     | VERSION |
|-------------|---------|
| store-redis | 3.4.0   |
| store-redis | 3.5.0   |
| mock-soap   | 3.5.1   |
| store-redis | 3.5.1   |

`

**Comment by bhavikas-inx** [comment]
Hi, I updated the imposter using Homebrew. Now I am getting 3.5.2 release.
Now the issue is changed. Now I am getting this for SOAP.

`root087@root087:~/BhavikaSoni/Projects/BeornTech/CigWorc/imposter$ imposter plugin install -d mock-soap
DEBU[0000] plugin mock-soap version 3.5.2 is not installed 
DEBU[0000] downloading https://github.com/outofcoffee/imposter/releases/download/v3.5.2/imposter-plugin-mock-soap.jar 
INFO[0001] downloaded plugin mock-soap version 3.5.2    
DEBU[0001] plugin mock-soap version 3.5.2 is installed  
ℹ️ Note that these plugins have not been saved as default plugins.
This means they are installed only for engine version 3.5.2 and not any future engine versions.
To change this behaviour, pass the --save-default (-d) flag.
INFO[0001] 1 plugin(s) installed                        
root087@root087:~/BhavikaSoni/Projects/BeornTech/CigWorc/imposter$ imposter up examples/soap/simple
DEBU[0000] plugin mock-soap version 3.5.2 is installed  
INFO[0000] starting mock engine on port 8080 - press ctrl+c to stop 
DEBU[0000] engine image '3.5.2' already present         
07:57:53 INFO  i.g.i.Imposter - Starting mock engine 3.5.2
07:57:53 DEBUG i.g.i.c.u.ConfigUtil - Loading configuration file: /opt/imposter/config/imposter-config.yaml
07:57:54 ERROR i.v.c.i.l.c.VertxIsolatedDeployer - Failed in deploying verticle
java.lang.RuntimeException: Could not find the plugin 'soap' - is it installed in the plugin path (/opt/imposter/plugins or program classpath)?
	at io.gatehill.imposter.plugin.DynamicPluginDiscoveryStrategyImpl.registerPluginClass(DynamicPluginDiscoveryStrategyImpl.kt:137) ~[imposter-config-dynamic-3.5.2.jar:?]
	at io.gatehill.imposter.plugin.DynamicPluginDiscoveryStrategyImpl.processPlugin(DynamicPluginDiscoveryStrategyImpl.kt:108) ~[imposter-config-dynamic-3.5.2.jar:?]
	at io.gatehill.imposter.plugin.DynamicPluginDiscoveryStrategyImpl.preparePluginsFromConfig(DynamicPluginDiscoveryStrategyImpl.kt:87) ~[imposter-config-dynamic-3.5.2.jar:?]
	at io.gatehill.imposter.plugin.PluginManagerImpl.preparePluginsFromConfig(PluginManagerImpl.kt:67) ~[imposter-config-3.5.2.jar:?]
	at io.gatehill.imposter.Imposter$start$1.invokeSuspend(Imposter.kt:114) ~[imposter-engine-3.5.2.jar:?]
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33) ~[kotlin-stdlib-1.7.10.jar:1.7.10-release-333(1.7.10)]
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106) ~[kotlinx-coroutines-core-jvm-1.6.4.jar:?]
	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:570) ~[kotlinx-coroutines-core-jvm-1.6.4.jar:?]
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:750) ~[kotlinx-coroutines-core-jvm-1.6.4.jar:?]
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:677) ~[kotlinx-coroutines-core-jvm-1.6.4.jar:?]
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:664) ~[kotlinx-coroutines-core-jvm-1.6.4.jar:?]
FATA[0030] timed out waiting for status endpoint to return HTTP 200 at http://localhost:8080/system/status `

**Comment by outofcoffee** [comment]
Hi @bhavikas-inx - this might be an issue with the path to the plugins and user permissions.

Would you mind running with the `--log-level trace` argument so it's possible to see the paths involved?

**Comment by merlinc** [comment]
I've got the same sort of issue when running one of the example apps. I've downloaded the imposter CLI binary, and put it in my path.

```bash
$ which imposter
/home/username/.local/bin/imposter

$ ls -1
imposter-config.yaml
petstore.wsdl
schema.xsd

$ imposter up --log-level trace
TRAC[0000] using docker library
TRAC[0000] resolving latest version (cache allowed: true)
TRAC[0000] ensured directory: /home/username/.imposter
TRAC[0000] ensured directory: /home/username/.imposter
TRAC[0000] latest version cached value: 3.8.0       
TRAC[0000] resolved latest version: 3.8.0               
TRAC[0000] found 1 default plugin(s): [mock-soap]
TRAC[0000] ensuring 1 plugins: [mock-soap]
TRAC[0000] ensured directory: /home/username/.imposter/plugins
TRAC[0000] ensured plugin directory: /home/username/.imposter/plugins/3.8.0
TRAC[0000] plugin mock-soap version 3.8.0 already exists at: /home/username/.imposter/plugins/3.8.0/imposter-plugin-mock-soap.jar
DEBU[0000] plugin mock-soap version 3.8.0 is installed  
TRAC[0000] using docker engine
INFO[0000] starting mock engine on port 8080 - press ctrl+c to stop
DEBU[0000] engine image '3.8.0' already present
TRAC[0000] containers matching labels: []
TRAC[0000] no existing containers found matching labels: map[io.gatehill.imposter.hash:01860b313e0929ca7dfc81118a8a21b1137ef083]
TRAC[0000] container user:
TRAC[0000] engine environment: [IMPOSTER_LOG_LEVEL=DEBUG IMPOSTER_CACHE_DIR=/tmp/imposter-cache IMPOSTER_OPENAPI_REMOTE_FILE_CACHE=true]
TRAC[0000] plugins are enabled
TRAC[0000] ensured directory: /home/username/.imposter/plugins
TRAC[0000] ensured plugin directory: /home/username/.imposter/plugins/3.8.0
TRAC[0000] file cache enabled                           
TRAC[0000] ensured directory: /home/username/.imposter/filecache
TRAC[0000] ensured file cache directory: /home/username/.imposter/filecache
TRAC[0000] using binds: [/home/username/imposter-demo/simple:/opt/imposter/config /home/username/.imposter/plugins/3.8.0:/opt/imposter/plugins /home/username/.imposter/filecache:/tmp/imposter-cache]                                                        
TRAC[0000] starting Docker mock engine              
TRAC[0000] waiting for status endpoint to return HTTP 200 at http://localhost:8080/system/status
13:50:31 INFO  i.g.i.Imposter - Starting mock engine 3.8.0
13:50:31 DEBUG i.g.i.c.u.ConfigUtil - Loading configuration file: /opt/imposter/config/imposter-config.yaml
13:50:31 ERROR i.v.c.i.l.c.VertxIsolatedDeployer - Failed in deploying verticle
java.lang.RuntimeException: Could not find the plugin 'soap' - is it installed in the plugin path (/opt/imposter/plugins or program classpath)?
        at io.gatehill.imposter.plugin.DynamicPluginDiscoveryStrategyImpl.registerPluginClass(DynamicPluginDiscoveryStrategyImpl.kt:137) ~[imposter-config-dynamic-3.8.0.jar:?]
        at io.gatehill.imposter.plugin.DynamicPluginDiscoveryStrategyImpl.processPlugin(DynamicPluginDiscoveryStrategyImpl.kt:108) ~[imposter-config-dynamic-3.8.0.jar:?]
        at io.gatehill.imposter.plugin.DynamicPluginDiscoveryStrategyImpl.preparePluginsFromConfig(DynamicPluginDiscoveryStrategyImpl.kt:87) ~[imposter-config-dynamic-3.8.0.jar:?]
        at io.gatehill.imposter.plugin.PluginManagerImpl.preparePluginsFromConfig(PluginManagerImpl.kt:67) ~[imposter-config-3.8.0.jar:?]
        at io.gatehill.imposter.Imposter$start$1.invokeSuspend(Imposter.kt:114) ~[imposter-engine-3.8.0.jar:?]
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33) ~[kotlin-stdlib-1.7.22.jar:1.7.22-release-288(1.7.22)]
        at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106) ~[kotlinx-coroutines-core-jvm-1.6.4.jar:?]
        at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:570) ~[kotlinx-coroutines-core-jvm-1.6.4.jar:?]
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:750) ~[kotlinx-coroutines-core-jvm-1.6.4.jar:?]
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:677) ~[kotlinx-coroutines-core-jvm-1.6.4.jar:?]
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:664) ~[kotlinx-coroutines-core-jvm-1.6.4.jar:?]
FATA[0030] timed out waiting for status endpoint to return HTTP 200 at http://localhost:8080/system/status
```

It looks like its finding it correctly, and the binding to the plugins directory looks ok, but it's still not being found at runtime.

**Comment by outofcoffee** [comment]
Thanks for the detailed report @merlinc 

Agree that the binding to the plugins dir looks correct:
```
TRAC[0000] using binds: [/home/username/imposter-demo/simple:/opt/imposter/config /home/username/.imposter/plugins/3.8.0:/opt/imposter/plugins /home/username/.imposter/filecache:/tmp/imposter-cache]                                                        
```

What is the output of doing:
```
ls -l /home/username/.imposter/plugins/3.8.0
```

Things it would be good to look for:
- are the plugin binaries there?
- do they match the files on https://github.com/outofcoffee/imposter/releases/tag/v3.8.0

A mitigation in the meantime would be to use the `all` distro, which bundles all the plugins like `soap` within it. I'll post some steps here.

**Comment by outofcoffee** [comment]
This should be fixed in [3.9.0](https://github.com/outofcoffee/imposter/releases/tag/v3.9.0). You shouldn't need to use the `docker-all` option from 3.9.0 onwards.
